aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -199,12 +209,17 @@ In apps that process requests, scoped services are disposed at the end of the re
 
 ### Singleton
 
-Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created the first time they're requested (or when `Startup.ConfigureServices` is run and an instance is specified with the service registration). Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.
+Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created either:
 
-In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed at app shutdown.
+* The first time they're requested.
+* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.
+<!-- REVIEW what is user code? -->
+Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.
```

Comment:
I think _user code_ here means implementing the Singleton pattern manually. e.g. using one of [Jon Skeet's patterns](https://csharpindepth.com/articles/singleton).

Reply:
I think "user code" in this case is referring code that might incorrectly manually dispose a resolved singleton.

Services should never be disposed by any code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton for you.

If an object instance is registered as a singleton, the code that created that instance is responsible for the lifetime, but the instance still must outlive the container. A reference to the singleton instance should be kept outside of the container. The singleton should then be disposed via this external reference after the container/host has already been disposed.